modules/distributor/forwarder.go

```diff
@@ -0,0 +1,311 @@
+package distributor
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/services"
+	"github.com/grafana/tempo/modules/overrides"
+	"github.com/grafana/tempo/pkg/util/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"go.uber.org/atomic"
+	"go.uber.org/multierr"
+)
+
+var (
+	metricForwarderPushes = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "tempo",
+		Name:      "distributor_forwarder_pushes",
+		Help:      "Total number of successful requests queued up for a tenant to the forwarder",
+	}, []string{"tenant"})
+	metricForwarderDroppedPushes = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "tempo",
+		Name:      "distributor_forwarder_dropped_pushes",
+		Help:      "Total number of failed pushes to the queue for a tenant to the forwarder",
+	}, []string{"tenant"})
+	metricForwarderQueueLength = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "tempo",
+		Name:      "distributor_forwarder_queue_length",
+		Help:      "Number of queued requests for a tenant",
+	}, []string{"tenant"})
+)
+
+type forwardFunc func(ctx context.Context, tenantID string, keys []uint32, traces []*rebatchedTrace) error
+
+type request struct {
+	keys   []uint32
+	traces []*rebatchedTrace
+}
+
+// forwarder queues up traces to be sent to the metrics-generators
+type forwarder struct {
+	services.Service
+
+	// per-tenant queue managers
+	queueManagers map[string]*queueManager
+	mutex         sync.Mutex
+
+	forwardFunc forwardFunc
+
+	o                 *overrides.Overrides
+	overridesInterval time.Duration
+	shutdown          chan interface{}
+}
+
+func newForwarder(fn forwardFunc, o *overrides.Overrides) *forwarder {
+	rf := &forwarder{
+		queueManagers:     make(map[string]*queueManager),
+		mutex:             sync.Mutex{},
+		forwardFunc:       fn,
+		o:                 o,
+		overridesInterval: time.Minute, // TODO: Make this configurable
+		shutdown:          make(chan interface{}),
+	}
+
+	rf.Service = services.NewIdleService(rf.start, rf.stop)
+
+	return rf
+}
+
+// SendTraces queues up traces to be sent to the metrics-generators
+func (f *forwarder) SendTraces(ctx context.Context, tenantID string, keys []uint32, traces []*rebatchedTrace) {
+	f.mutex.Lock()
+	defer f.mutex.Unlock()
```

Comment:
Yeah, I refrained myself of using a `sync.RWLock` to avoid the hassle of changing locks. I wanted to test this first with high throughputs and see if there was lock contention. I'm probably just being too lazy 😅